syncstorage/src/web/extractors.rs

```diff
@@ -374,94 +374,95 @@ impl FromRequest for BsoBody {
     type Future = LocalBoxFuture<'static, Result<BsoBody, Self::Error>>;
 
     fn from_request(req: &HttpRequest, payload: &mut Payload) -> Self::Future {
-        // Only try and parse the body if its a valid content-type
-        let ctype = match ContentType::parse(req) {
-            Ok(v) => v,
-            Err(e) => {
-                return Box::pin(future::err(
-                    ValidationErrorKind::FromDetails(
+        // req.clone() allows move into async block since it is borrowed
+        // payload.take() grabs request body payload, replacing the one passed in
+        // with an empty payload so we strictly read the request body payload once
+        // and dispense with it
+        let req = req.clone();
+        let mut payload = payload.take();
+
+        Box::pin(async move {
+            // Only try and parse the body if its a valid content-type
+            let ctype = match ContentType::parse(&req) {
```

Comment:
This could further refactor into `map_err(..)?` and similarly below w/ the `req.app_data` (which could use `ok_or_else` since it's an `Option`)